### DISCUSSION TITLE
WASM asset loading best practices

### DISCUSSION THREAD

**Discussion by 3elDU** [hajimehoshi/ebiten#2656] [Q&A] [answered]
When compiling the project for wasm, the filesystem is not available, since we're running in a browser. So how do you load assets? What are the best practices?
My first thought was to check if we are running in wasm, and in that case, load assets with http requests. For that we need to host those assets somewhere first, of course.
Maybe assets could be embedded in the wasm executable?
What are your thoughts?

**Comment by tinne26** [comment, marked as answer]
The most common approach is to use the <https://pkg.go.dev/embed> package in order to embed the assets directly into the compiled executable / wasm file. Most entries in the 2022 Ebitengine [game jam](https://itch.io/jam/ebiten-game-jam/results) did that. If you have many assets, though, yes, you would need to use http requests, but that's rarely necessary. Like, if you need to do that, you should be at a level where you can figure it out on your own.

**Reply by 3elDU** [reply]
That worked fine! It's great I can even embed a whole directory.
The other issue I encountered though is using `indexedDB` from Go for saving/loading mechanic. Or speaking more generally, calling async JS functions from Go. I tried a lot of ways, and I either get a deadlock, or nothing happens at all.

**Reply by tinne26** [reply]
Honestly, using js from Go/wasm with syscall/js is a mess. It can be done, but it's not simple, e.g.: <https://dev.to/clavinjune/golang-wasm-async-function-35j2>. But this is really worth of a separate package, e.g. <https://github.com/hack-pad/go-indexeddb>.

**Reply by 3elDU** [reply]
This looks like just what I wanted! Thanks, I'll take a look.
